File: substrate/recursive_twin_search_prompt_context_compose.py
"""Recursive twin search → prompt context pack (pure).

twin_written, remote_index_queried, record_persisted, prompts_injected,
live_router_authorized always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.recursive_twin_intelligent_search import (
    TwinIntelligentSearchError,
    TwinSearchResult,
    search_twin_substrate,
)
from substrate.recursive_twin_note_taker_compose import (
    RecursiveTwinNoteTakerCompose,
    RecursiveTwinNoteTakerComposeError,
    compose_recursive_twin_note_taker,
)
from substrate.workstation_record_prompt_model_decision_compose import (
    WorkstationRecordPromptModelDecisionCompose,
    WorkstationRecordPromptModelDecisionComposeError,
    compose_workstation_record_prompt_model_decision,
)


class RecursiveTwinSearchPromptContextComposeError(ValueError):
    """Fail-closed validation for twin search prompt context pack."""


@dataclass(frozen=True)
class RecursiveTwinSearchPromptContextCompose:
    session_id: str
    parent_asset_id: str
    twin_propose: RecursiveTwinNoteTakerCompose
    search: TwinSearchResult
    prompt_pack: WorkstationRecordPromptModelDecisionCompose
    pack_ready: bool
    twin_written: bool
    remote_index_queried: bool
    record_persisted: bool
    prompts_injected: bool
    live_router_authorized: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "twin_propose": self.twin_propose.to_dict(),
            "search": self.search.to_dict(),
            "prompt_pack": self.prompt_pack.to_dict(),
            "pack_ready": self.pack_ready,
            "twin_written": False,
            "remote_index_queried": False,
            "record_persisted": False,
            "prompts_injected": False,
            "live_router_authorized": False,
            "notes": list(self.notes),
            "authority": (
                "recursive_twin_search_prompt_context_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise RecursiveTwinSearchPromptContextComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _hits_to_session_records(search: TwinSearchResult) -> list[dict[str, Any]]:
    records: list[dict[str, Any]] = []
    for hit in search.hits:
        for i, snip in enumerate(hit.snippets):
            is_question = (
                "questions" in hit.matched_fields
                and ("?" in snip or hit.matched_fields[0] == "questions")
            )
            records.append(
                {
                    "record_id": f"{hit.twin_id}-s{i}",
                    "kind": "question" if is_question else "insight",
                    "body": snip,
                    "source_ref": hit.parent_asset_id,
                }
            )
    return records


def compose_recursive_twin_search_prompt_context(
    *,
    session_id: object,
    parent_asset_id: object,
    source_excerpt: object,
    twin_records: object,
    search_query: object,
    user_prompt: object,
    selected_model_id: object,
    models: object,
    daily_cap_usd: object,
    spent_usd: object,
    operator_ack: object,
    existing_twin_asset_id: object | None = None,
    focus_questions: object | None = None,
    search_limit: object | None = None,
    projected_cost_usd_high: object | None = None,
    projected_cost_usd_low: object | None = None,
    bench_bests: object | None = None,
    focus_task: object | None = None,
    nd_shadow: object | None = None,
) -> RecursiveTwinSearchPromptContextCompose:
    """Twin propose + search + prompt pack. Never writes/injects/routes."""
    if not isinstance(operator_ack, bool):
        raise RecursiveTwinSearchPromptContextComposeError(
            "operator_ack must be an explicit boolean"
        )
    session = _require_nonempty(session_id, field="session_id")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")

    notes: list[str] = [
        "twin_written=false — twin note-taker is propose/scaffold only",
        "remote_index_queried=false — twin search is pure local corpus",
        "record_persisted=false · prompts_injected=false · live_router_authorized=false",
    ]

    try:
        twin_propose = compose_recursive_twin_note_taker(
            parent_asset_id=parent,
            source_excerpt=source_excerpt,
            operator_ack=operator_ack,
            existing_twin_asset_id=existing_twin_asset_id,
            focus_questions=focus_questions,
        )
    except RecursiveTwinNoteTakerComposeError as e:
        raise RecursiveTwinSearchPromptContextComposeError(str(e)) from e
    notes.extend(f"[twin] {n}" for n in twin_propose.notes)

    if not isinstance(twin_records, list):
        raise RecursiveTwinSearchPromptContextComposeError(
            "twin_records must be an array"
        )
    lim = 20 if search_limit is None else search_limit
    try:
        search = search_twin_substrate(
            query=search_query,
            records=twin_records,
            limit=lim,
        )
    except TwinIntelligentSearchError as e:
        raise RecursiveTwinSearchPromptContextComposeError(str(e)) from e
    notes.extend(f"[search] {n}" for n in search.notes)
    notes.append(f"search_hits={len(search.hits)}")

    session_records = _hits_to_session_records(search)
    if len(session_records) == 0 and operator_ack:
        notes.append(
            "no twin search hits — seed scaffold insight from source_excerpt "
            "(caller text only)"
        )
        excerpt = _require_nonempty(source_excerpt, field="source_excerpt")
        session_records.append(
            {
                "record_id": "scaffold-excerpt",
                "kind": "insight",
                "body": excerpt[:500],
                "source_ref": parent,
            }
        )

    try:
        prompt_pack = compose_workstation_record_prompt_model_decision(
            session_id=session,
            parent_asset_id=parent,
            records=session_records,
            user_prompt=user_prompt,
            selected_model_id=selected_model_id,
            models=models,
            daily_cap_usd=daily_cap_usd,
            spent_usd=spent_usd,
            operator_ack=operator_ack,
            projected_cost_usd_high=projected_cost_usd_high,
            projected_cost_usd_low=projected_cost_usd_low,
            bench_bests=bench_bests,
            focus_task=focus_task,
            nd_shadow=nd_shadow,
        )
    except WorkstationRecordPromptModelDecisionComposeError as e:
        raise RecursiveTwinSearchPromptContextComposeError(str(e)) from e
    notes.extend(f"[prompt] {n}" for n in prompt_pack.notes)

    pack_ready = (
        twin_propose.twin_propose_ready is True
        and prompt_pack.pack_ready is True
        and operator_ack is True
    )
    if pack_ready:
        notes.append(
            "pack_ready=true — twin propose + search + prompt context advisory pack"
        )
    else:
        notes.append(
            "pack_ready=false — twin, search, prompt pack, or operator_ack gate open"
        )

    if (
        twin_propose.twin_written is not False
        or twin_propose.prompts_injected is not False
        or search.remote_index_queried is not False
        or prompt_pack.record_persisted is not False
        or prompt_pack.prompts_injected is not False
        or prompt_pack.live_router_authorized is not False
    ):
        raise RecursiveTwinSearchPromptContextComposeError(
            "invariant: honesty flags must remain false"
        )

    notes.extend(
        (
            "twin_written=false",
            "remote_index_queried=false",
            "record_persisted=false",
            "prompts_injected=false",
            "live_router_authorized=false",
        )
    )

    return RecursiveTwinSearchPromptContextCompose(
        session_id=session,
        parent_asset_id=parent,
        twin_propose=twin_propose,
        search=search,
        prompt_pack=prompt_pack,
        pack_ready=pack_ready,
        twin_written=False,
        remote_index_queried=False,
        record_persisted=False,
        prompts_injected=False,
        live_router_authorized=False,
        notes=tuple(notes),
        authority="recursive_twin_search_prompt_context_compose_advisory",
    )


def format_recursive_twin_search_prompt_context_summary(
    c: RecursiveTwinSearchPromptContextCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · hits={len(c.search.hits)} · "
        f"twin_propose_ready={c.twin_propose.twin_propose_ready} · "
        f"would_exceed={c.prompt_pack.would_exceed} · "
        f"twin_written=false · remote_index_queried=false · prompts_injected=false"
    )


__all__ = [
    "RecursiveTwinSearchPromptContextCompose",
    "RecursiveTwinSearchPromptContextComposeError",
    "compose_recursive_twin_search_prompt_context",
    "format_recursive_twin_search_prompt_context_summary",
]

File: apps/reading/src/api/recursiveTwinSearchPromptContextCompose.ts
/**
 * Recursive twin note-taker → intelligent search → prompt context pack (pure).
 *
 * Operator vision: every information asset has a twin substrate of
 * insights/questions; search that substrate and leverage hits into prompt
 * context for the next research/write step — without inventing notes or
 * live-injecting prompts.
 *
 * twin_written always false.
 * remote_index_queried always false.
 * record_persisted always false.
 * prompts_injected always false.
 * live_router_authorized always false.
 */

import {
  composeRecursiveTwinNoteTaker,
  type RecursiveTwinNoteTakerCompose,
} from "./recursiveTwinNoteTakerCompose";
import {
  searchTwinSubstrate,
  type TwinSearchRecord,
  type TwinSearchResult,
} from "./recursiveTwinIntelligentSearch";
import {
  composeWorkstationRecordPromptModelDecision,
  type WorkstationRecordPromptModelDecisionCompose,
} from "./workstationRecordPromptModelDecisionCompose";
import type { SessionRecordItem } from "./workstationSessionInsightRecordCompose";
import type { ModelOption } from "./modelDecisionPromptCompose";
import type { BenchTaskBest } from "./settingsModelDriverTabCompose";
import type { NotDiamondShadowRec } from "./settingsModelDriverTabCompose";

export interface RecursiveTwinSearchPromptContextInput {
  session_id: string;
  parent_asset_id: string;
  /** Source excerpt for twin note-taker scaffold (caller-supplied). */
  source_excerpt: string;
  existing_twin_asset_id?: string | null;
  focus_questions?: string[] | null;
  /** Twin substrate corpus to search (caller-supplied; never invented). */
  twin_records: TwinSearchRecord[];
  search_query: string;
  search_limit?: number;
  user_prompt: string;
  selected_model_id: string;
  models: ModelOption[];
  daily_cap_usd: number | null;
  spent_usd: number | null;
  projected_cost_usd_high?: number | null;
  projected_cost_usd_low?: number | null;
  bench_bests?: BenchTaskBest[] | null;
  focus_task?: string | null;
  nd_shadow?: NotDiamondShadowRec | null;
  operator_ack: boolean;
}

export interface RecursiveTwinSearchPromptContextCompose {
  session_id: string;
  parent_asset_id: string;
  twin_propose: RecursiveTwinNoteTakerCompose;
  search: TwinSearchResult;
  prompt_pack: WorkstationRecordPromptModelDecisionCompose;
  /**
   * True when twin_propose_ready (or existing twin), search composed,
   * and prompt_pack.pack_ready. Still never writes/injects/routes.
   */
  pack_ready: boolean;
  twin_written: false;
  remote_index_queried: false;
  record_persisted: false;
  prompts_injected: false;
  live_router_authorized: false;
  notes: string[];
  authority: "recursive_twin_search_prompt_context_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Map twin search hits → session records for prompt context bridge.
 * Snippets are caller-derived from corpus only.
 */
function hitsToSessionRecords(
  search: TwinSearchResult,
): SessionRecordItem[] {
  const records: SessionRecordItem[] = [];
  for (const hit of search.hits) {
    for (let i = 0; i < hit.snippets.length; i++) {
      const snip = hit.snippets[i];
      const isQuestion =
        hit.matched_fields.includes("questions") &&
        (snip.includes("?") || hit.matched_fields[0] === "questions");
      records.push({
        record_id: `${hit.twin_id}-s${i}`,
        kind: isQuestion ? "question" : "insight",
        body: snip,
        source_ref: hit.parent_asset_id,
      });
    }
  }
  return records;
}

/**
 * Twin propose + substrate search + prompt/model decision pack.
 * Never writes twins; never remote-indexes; never injects prompts.
 */
export function composeRecursiveTwinSearchPromptContext(
  input: RecursiveTwinSearchPromptContextInput,
): RecursiveTwinSearchPromptContextCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );

  const notes: string[] = [
    "twin_written=false — twin note-taker is propose/scaffold only",
    "remote_index_queried=false — twin search is pure local corpus",
    "record_persisted=false · prompts_injected=false · live_router_authorized=false",
  ];

  const twin_propose = composeRecursiveTwinNoteTaker({
    parent_asset_id,
    source_excerpt: input.source_excerpt,
    existing_twin_asset_id: input.existing_twin_asset_id,
    operator_ack: input.operator_ack,
    focus_questions: input.focus_questions,
  });
  notes.push(...twin_propose.notes.map((n) => `[twin] ${n}`));

  if (!Array.isArray(input.twin_records)) {
    throw new Error("twin_records must be an array");
  }
  const search = searchTwinSubstrate({
    query: input.search_query,
    records: input.twin_records,
    limit: input.search_limit,
  });
  notes.push(...search.notes.map((n) => `[search] ${n}`));
  notes.push(`search_hits=${search.hits.length}`);

  const sessionRecords = hitsToSessionRecords(search);
  // If no hits, still allow empty records path? workstation may require records —
  // seed one insight from focus_questions or source excerpt marker when ack.
  if (sessionRecords.length === 0 && input.operator_ack) {
    notes.push(
      "no twin search hits — seed scaffold insight from source_excerpt (caller text only)",
    );
    sessionRecords.push({
      record_id: "scaffold-excerpt",
      kind: "insight",
      body: input.source_excerpt.trim().slice(0, 500),
      source_ref: parent_asset_id,
    });
  }

  const prompt_pack = composeWorkstationRecordPromptModelDecision({
    session_id,
    parent_asset_id,
    records: sessionRecords,
    user_prompt: input.user_prompt,
    selected_model_id: input.selected_model_id,
    models: input.models,
    daily_cap_usd: input.daily_cap_usd,
    spent_usd: input.spent_usd,
    projected_cost_usd_high: input.projected_cost_usd_high,
    projected_cost_usd_low: input.projected_cost_usd_low,
    bench_bests: input.bench_bests,
    focus_task: input.focus_task,
    nd_shadow: input.nd_shadow,
    operator_ack: input.operator_ack,
  });
  notes.push(...prompt_pack.notes.map((n) => `[prompt] ${n}`));

  const pack_ready =
    twin_propose.twin_propose_ready === true &&
    prompt_pack.pack_ready === true &&
    input.operator_ack === true;

  if (pack_ready) {
    notes.push(
      "pack_ready=true — twin propose + search + prompt context advisory pack",
    );
  } else {
    notes.push(
      "pack_ready=false — twin, search, prompt pack, or operator_ack gate open",
    );
  }

  if (
    twin_propose.twin_written !== false ||
    twin_propose.prompts_injected !== false ||
    search.remote_index_queried !== false ||
    prompt_pack.record_persisted !== false ||
    prompt_pack.prompts_injected !== false ||
    prompt_pack.live_router_authorized !== false
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("twin_written=false");
  notes.push("remote_index_queried=false");
  notes.push("record_persisted=false");
  notes.push("prompts_injected=false");
  notes.push("live_router_authorized=false");

  return {
    session_id,
    parent_asset_id,
    twin_propose,
    search,
    prompt_pack,
    pack_ready,
    twin_written: false,
    remote_index_queried: false,
    record_persisted: false,
    prompts_injected: false,
    live_router_authorized: false,
    notes,
    authority: "recursive_twin_search_prompt_context_compose_advisory",
  };
}

export function formatRecursiveTwinSearchPromptContextSummary(
  c: RecursiveTwinSearchPromptContextCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · hits=${c.search.hits.length} · ` +
    `twin_propose_ready=${c.twin_propose.twin_propose_ready} · ` +
    `would_exceed=${c.prompt_pack.would_exceed} · ` +
    `twin_written=false · remote_index_queried=false · prompts_injected=false`
  );
}
